nvbench/option_parser.cu

```diff
@@ -781,6 +782,8 @@ catch (std::exception &e)
 
 void option_parser::replay_global_args()
 {
+  // Initialize benchmark with all devices:
+  m_benchmarks.back()->set_devices(nvbench::device_manager::get().get_devices());
```

Comment:
important: it seems the contract is that devices are always set before benchmark is used. It took me a while to track setting and getting devices in current codebase to see if this change is safe.  From the language perspective, nothing guarantees that devices are set before they are used. I think this makes code fragile. I'm not sure how the right refactoring would look like, but we could at least use optional:
```
std::optional<std::vector<nvbench::device_info>> m_devices;
```
This way we'll get an exception if devices are referenced before they are set.

Reply:
I agree that this is fragile, but I'd rather not have an empty device list be an error condition. If a user programmatically filters devices out and ends up with none left in a particular benchmark, skipping the benchmark instead of throwing an error seems least surprising IMO.

I think I see a clean way of doing this without being too invasive. Since the benchmarks are always cloned from the `benchmark_manager`, we can setup benchmark defaults in a `benchmark_manager::initialize` function from `main`. Then the devices will get cloned with the full set of default devices, before the `option_parser` potentially reduces the coverage.